contracts/src/RIP7755Source.sol

```diff
@@ -0,0 +1,271 @@
+    uint256 private _nonce;
+
+    /// @notice Event emitted when a user requests a cross chain call to be made by a filler
+    /// @param requestHash The keccak256 hash of a `CrossChainRequest`
+    /// @param request The requested cross chain call
+    event CrossChainCallRequested(bytes32 indexed requestHash, CrossChainRequest request);
+
+    /// @notice Event emitted when an expired cross chain call request is canceled
+    /// @param requestHash The keccak256 hash of a `CrossChainRequest`
+    event CrossChainCallCanceled(bytes32 indexed requestHash);
+
+    /// @notice This error is thrown when a cross chain request specifies the native currency as the reward type but
+    /// does not send the correct `msg.value`
+    /// @param expected The expected `msg.value` that should have been sent with the transaction
+    /// @param received The actual `msg.value` that was sent with the transaction
+    error InvalidValue(uint256 expected, uint256 received);
+
+    /// @notice This error is thrown if a user attempts to cancel a request that is not in the `CrossChainCallStatus.Requested` state
+    /// @param status The `CrossChainCallStatus` status that the request has
+    error InvalidStatusForRequestCancel(CrossChainCallStatus status);
+
+    /// @notice This error is thrown if a Filler attempts to claim a reward for a request that is not in a `CrossChainCallStatus.Requested` state
+    /// @param status The `CrossChainCallStatus` status of the request
+    error InvalidStatusForClaim(CrossChainCallStatus status);
+
+    /// @notice This error is thrown if an attempt to cancel a request is made before the request's expiry timestamp
+    /// @param currentTimestamp The current block timestamp
+    /// @param expiry The timestamp at which the request expires
+    error CannotCancelRequestBeforeExpiry(uint256 currentTimestamp, uint256 expiry);
+
+    /// @notice Submits an RIP-7755 request for a cross chain call
+    ///
+    /// @param request A cross chain request structured as a `CrossChainRequest`
+    function requestCrossChainCall(CrossChainRequest memory request) external payable {
+        request.nonce = ++_nonce;
+        bool usingNativeCurrency = request.rewardAsset == _NATIVE_ASSET;
+        uint256 expectedValue = usingNativeCurrency ? request.rewardAmount : 0;
+
+        if (msg.value != expectedValue) {
+            revert InvalidValue(expectedValue, msg.value);
+        }
+
+        bytes32 requestHash = hashRequestMemory(request);
+        _requestMetadata[requestHash] = RequestMeta({
+            status: CrossChainCallStatus.Requested,
+            expiryTimestamp: uint40(block.timestamp + request.validDuration),
+            requester: msg.sender
+        });
+
+        emit CrossChainCallRequested(requestHash, request);
+
+        if (!usingNativeCurrency) {
+            _pullERC20({owner: msg.sender, asset: request.rewardAsset, amount: request.rewardAmount});
```

Comment:
just curious, why do this last, instead of just after the `msg.value` check above?

Reply:
I guess this doesn't have to be last. Just force of habit to prevent reentrancies since it's calling to an unknown contract address 😅. Do you have a preference to move it up with the `msg.value` check?

Reply:
generally think events should come last, but good to have external calls after state changes!